api/clients/v2/eigenda_client.go

```diff
@@ -0,0 +1,279 @@
+	relayClient RelayClient,
+	blobVerifier verification.IBlobVerifier,
+	codec codecs.BlobCodec,
+	g1Srs []bn254.G1Affine) (*EigenDAClient, error) {
+
+	return &EigenDAClient{
+		log:          log,
+		random:       random,
+		clientConfig: clientConfig,
+		codec:        codec,
+		relayClient:  relayClient,
+		blobVerifier: blobVerifier,
+		g1Srs:        g1Srs,
+	}, nil
+}
+
+// GetPayload iteratively attempts to fetch a given blob with key blobKey from relays that have it, as claimed by the
+// blob certificate. The relays are attempted in random order.
+//
+// If the blob is successfully retrieved, then the blob is verified. If the verification succeeds, the blob is decoded
+// to yield the payload (the original user data), and the payload is returned.
+func (c *EigenDAClient) GetPayload(
+	ctx context.Context,
+	blobKey core.BlobKey,
+	eigenDACert *verification.EigenDACert) ([]byte, error) {
+
+	err := c.verifyCertWithTimeout(ctx, eigenDACert)
+	if err != nil {
+		return nil, fmt.Errorf("verify cert with timeout for blobKey %v: %w", blobKey, err)
+	}
+
+	relayKeys := eigenDACert.BlobVerificationProof.BlobCertificate.RelayKeys
+	relayKeyCount := len(relayKeys)
+	if relayKeyCount == 0 {
+		return nil, errors.New("relay key count is zero")
+	}
+
+	blobCommitmentProto := contractEigenDABlobVerifier.BlobCommitmentBindingToProto(
+		&eigenDACert.BlobVerificationProof.BlobCertificate.BlobHeader.Commitment)
+	blobCommitment, err := encoding.BlobCommitmentsFromProtobuf(blobCommitmentProto)
+
+	if err != nil {
+		return nil, fmt.Errorf("blob commitments from protobuf: %w", err)
+	}
+
+	// create a randomized array of indices, so that it isn't always the first relay in the list which gets hit
+	indices := c.random.Perm(relayKeyCount)
+
+	// TODO (litt3): consider creating a utility which can deprioritize relays that fail to respond (or respond maliciously)
+
+	// iterate over relays in random order, until we are able to get the blob from someone
+	for _, val := range indices {
+		relayKey := relayKeys[val]
+
+		blob, err := c.getBlobWithTimeout(ctx, relayKey, blobKey)
+		// if GetBlob returned an error, try calling a different relay
+		if err != nil {
+			c.log.Warn("blob couldn't be retrieved from relay", "blobKey", blobKey, "relayKey", relayKey, "error", err)
+			continue
+		}
+
+		err = c.verifyBlobAgainstCert(blobKey, relayKey, blob, blobCommitment.Commitment, blobCommitment.Length)
+
+		// An honest relay should never send a blob which doesn't verify
+		if err != nil {
+			c.log.Warn("verify blob from relay: %w", err)
+			continue
+		}
+
+		payload, err := c.codec.DecodeBlob(blob)
+		if err != nil {
+			c.log.Error(
+				`Blob verification was successful, but decode blob failed!
+					This is likely a problem with the local blob codec configuration,
+					but could potentially indicate a maliciously generated blob certificate.
+					It should not be possible for an honestly generated certificate to verify
+					for an invalid blob!`,
```

Comment:
Think this is not 100% accurate. the cert only talks about the blob. Here its the decoding of the blob to the payload that is failing. Our disperser client would only construct correctly encoded blobs, but it might have been constructed by some other library (lets say our rust library), and there might be some incompatibility between the two for eg.

Reply:
To me, "honestly generated certificate" *includes* the dispersing client verifying that the input bytes have a valid encoding for the intended use. By "should not be possible", I'm trying to convey that this should never happen in the normal course of business. If it does happen, it indicates when of the following:

- a bug somewhere in the system
- a malicious dispersing client
- broken cryptography

Do you have any suggestions for rewording here?

Reply:
Not sure because I haven't wrapped my head around this whole length proof verification stuff, and the kind of attacks possible. But your "a malicious dispersing client" is PRECISELY the kind of thing EigenDA is meant to protect against (otherwise... just use S3). So we absolutely need good safeguards and guarantees around these attacks.

Reply:
Let's chat during our Integrations Food for Thought this afternoon. I'm fine in the meantime with merging this PR as is.

Reply:
> But your "a malicious dispersing client" is PRECISELY the kind of thing EigenDA is meant to protect against

I don't think it is. EigenDA is meant to guarantee that data is immutable and available. If the dispersing client wants to fill a blob with total gobbldygook that means nothing to anyone, DA isn't in a position to protect against that. All DA can do it guarantee that the gobbldygook is faithfully preserved